onnxruntime/core/providers/cpu/controlflow/loop.cc

```diff
@@ -463,9 +491,31 @@ Status LoopImpl::Execute(const FeedsFetchesManager& ffm) {
   // As the loop carried variables may change shape across iterations there's no way to avoid a copy
   // as we need the final shape.
   auto copy_tensor_from_mlvalue_to_output = [this](const OrtValue& input, int output_idx) {
-    auto& data = input.Get<Tensor>();
-    Tensor* output = context_.Output(output_idx, data.Shape());
-    session_state_.GetDataTransferMgr().CopyTensor(input.Get<Tensor>(), *output);
+    auto type = input.Type();
+    if (type == DataTypeImpl::GetType<Tensor>()) {
+      auto& data = input.Get<Tensor>();
+      Tensor* output = context_.Output(output_idx, data.Shape());
+      session_state_.GetDataTransferMgr().CopyTensor(input.Get<Tensor>(), *output);
+    } else if (type == DataTypeImpl::GetType<TensorSeq>()) {
+      std::vector<Tensor> tensors;
+
+      auto& data = input.Get<TensorSeq>();
+      TensorSeq* output = context_.Output<TensorSeq>(output_idx);
+      output->SetType(data.DataType());
+
+      AllocatorPtr alloc;
+      auto status = context_.GetTempSpaceAllocator(&alloc);
+      if (!status.IsOK()) {
+        ORT_THROW("Unable to get an allocator");
+      }
+      for (auto it = data.begin(); it != data.end(); ++it) {
+        Tensor tmp(it->DataType(), onnxruntime::TensorShape(it->Shape()), alloc);
```

Comment:
nit:call data.end() once
`for (auto it = data.begin(), end = data.end(); it != end; ++it){`